### PR TITLE
fix(route): 重新适配了CUG的各类网站-网站改版

### PR DESCRIPTION
**Pull request by Doradx**
## 该 PR 相关 Issue / Involved Issue

Close #

## 路由地址示例 / Example for the Proposed Route(s)

```routes
/cug/news
/cug/graduate
/cug/gcxy/0
/cug/gcxy/1
/cug/gcxy/2
/cug/gcxy/3
/cug/gcxy/4
/cug/gcxy/5
/cug/gcxy/6
```

## 新 RSS 路由检查表 / New RSS Route Checklist
  
- [ ] 新的路由 New Route
  - [ ] 跟随 [v2 路由规范](https://docs.rsshub.app/joinus/script-standard.html) Follows [v2 Script Standard](https://docs.rsshub.app/en/joinus/script-standard.html)
- [x] 文档说明 Documentation
  - [x] 中文文档 CN
  - [ ] 英文文档 EN
- [x] 全文获取 fulltext
  - [x] 使用缓存 Use Cache
- [ ] 反爬/频率限制 anti-bot or rate limit?
  - [ ] 如果有, 是否有对应的措施? If yes, do your code reflect this sign?
- [x] [日期和时间](https://docs.rsshub.app/joinus/pub-date.html) [date and time](https://docs.rsshub.app/en/joinus/pub-date.html)
  - [x] 可以解析 Parsed
  - [ ] 时区调整 Correct TimeZone
- [ ] 添加了新的包 New package added
- [ ] `Puppeteer`

## 说明 / Note
由于网站改版，之前的规则无法适应，故重新适配，并优化代码。

### CODE REVIEW

File: lib/routes/universities/cug/gcxy/index.js [hunk 1]
```diff
@@ -18,48 +19,40 @@ module.exports = async (ctx) => {
...
-    if (type === undefined) {
+    // check if the type is valid(is number and within range from 0 to typeUrl.length), if not, set it to 0
+    if (type === undefined || isNaN(type) || type < 0 || type >= typeUrl.length) {
         type = 0;
     } else {
-        type = Number(type);
-        if (type >= typeUrl.length) {
-            return false;
-        }
+        type = parseInt(type);
```
Comment (TonyRL) [inline lib/routes/universities/cug/gcxy/index.js:44]: The type of path variable will be either `String` or `undefined`. It won't be a `Number`.
docs: https://docs.rsshub.app/joinus/new-rss/before-start.html

File: lib/routes/universities/cug/gcxy/index.js [hunk 2]
```diff
@@ -70,40 +63,40 @@ module.exports = async (ctx) => {
...
-            const items = $('div.right_list > ul > li');
+            const rowList = $('ul.col-news-list > li.list_item');
+            // get all items from the rowList
+            const items = rowList
+                .map((index, item) => {
+                    const li = $(item);
+                    return {
+                        title: li.find('a.news-title').text(),
+                        link: new URL(li.find('a.news-title').attr('href'), host).href,
+                        description: '',
```
Comment (TonyRL) [inline lib/routes/universities/cug/gcxy/index.js:74]:
```suggestion
```

File: lib/routes/universities/cug/gcxy/index.js [hunk 3]
```diff
@@ -70,40 +63,40 @@ module.exports = async (ctx) => {
...
-                                    const $ = cheerio.load(res.data);
-                                    return $('.v_news_content').html();
-                                })
-                                .catch(() => ''),
-                            pubDate: item.find('span').text(),
-                        };
+                items.map((element) =>
+                    ctx.cache.tryGet(`cug/${element.link}`, async () => {
+                        try {
+                            const result = await got.get(String(element.link));
```
Comment (TonyRL) [inline lib/routes/universities/cug/gcxy/index.js:84]: https://github.com/DIYgod/RSSHub/blob/4fea5c788a5bf1c63f97122e930cd463421c87ee/lib/routes/universities/cug/gcxy/index.js#L73
`.href` returns a `String`. Don't recast the same type.
docs: https://nodejs.org/api/url.html#urlhref
https://docs.rsshub.app/joinus/script-standard.html

Comment (TonyRL) [inline lib/routes/universities/cug/gcxy/index.js:84]: Please stick with one convention only.
Either
https://github.com/DIYgod/RSSHub/blob/4fea5c788a5bf1c63f97122e930cd463421c87ee/lib/routes/universities/cug/gcxy/index.js#L58-L64
or
https://github.com/DIYgod/RSSHub/blob/4fea5c788a5bf1c63f97122e930cd463421c87ee/lib/routes/universities/cug/gcxy/index.js#L84

File: docs/university.md [hunk 4]
```diff
@@ -3597,11 +3597,11 @@ jsjxy.hbut.edu.cn 证书链不全，自建 RSSHub 可设置环境变量 NODE_TLS
 
 ### 工程学院
 
-<Route author="Dorad" example="/cug/gcxy/1" path="/cug/gcxy/:type" />
+<Route author="Dorad" example="/cug/gcxy/0" path="/cug/gcxy/:type" />
```
Comment (TonyRL) [inline docs/university.md:3600]: Since the previous example works, I don't see the necessity of changing this.

**Review by TonyRL — COMMENTED** [review]
https://github.com/DIYgod/RSSHub/pull/12739#discussion_r1251287700
https://github.com/DIYgod/RSSHub/pull/12739#discussion_r1251289177
https://github.com/DIYgod/RSSHub/pull/12739#discussion_r1252054154
https://github.com/DIYgod/RSSHub/pull/12739#discussion_r1252063672
https://github.com/DIYgod/RSSHub/pull/12739#discussion_r1252088515
https://github.com/DIYgod/RSSHub/pull/12739#discussion_r1252090801
https://github.com/DIYgod/RSSHub/pull/12739#discussion_r1252094271

**Review by Doradx — COMMENTED** [review]
all comments have been reviewed and resolved.

File: lib/routes/universities/cug/gcxy/index.js (hunk 2 above)
Comment (TonyRL) [reply lib/routes/universities/cug/gcxy/index.js:72]: Please explain how setting an empty property `description: ''` can help converting relative path to absolute path.

File: lib/routes/universities/cug/gcxy/index.js (hunk 3 above)
Comment (TonyRL) [reply lib/routes/universities/cug/gcxy/index.js:82]: Please explain how recasting `element.link` **from `string` to `string`** can help converting relative path to absolute path.

File: lib/routes/universities/cug/graduate.js [hunk 5]
```diff
@@ -1,38 +1,45 @@
...
-                };
+
+    const out = await Promise.all(
+        items.map((element) =>
+            ctx.cache.tryGet(`cug/${element.link}`, async () => {
+                try {
+                    if (!element.link.includes('graduate.cug.edu.cn')) {
+                        return element;
+                    }
+                    const result = await got.get(element.link);
```
[the remark this answers is not in the document]
Comment (TonyRL) [reply lib/routes/universities/cug/graduate.js:32]: This is not fixed at all.

File: lib/routes/universities/cug/news.js [hunk 6]
```diff
@@ -37,27 +39,25 @@ module.exports = async (ctx) => {
...
-            } catch (e) {
-                element.enabled = false;
-            }
-            return Promise.resolve(element);
-        })
+    const out = await Promise.all(
+        results.map((element) =>
+            ctx.cache.tryGet(`cug/${element.link}`, async () => {
+                try {
+                    const result = await got.get(element.link);
```
[the remark this answers is not in the document]
Comment (TonyRL) [reply lib/routes/universities/cug/news.js:50]: This is not fixed at all.

**Review by TonyRL — CHANGES_REQUESTED** [review]

**Pull request closed without merging**